test/integration/utility.cc

```diff
@@ -75,30 +83,59 @@ void BufferingStreamDecoder::onResetStream(Http::StreamResetReason, absl::string
   ADD_FAILURE();
 }
 
-BufferingStreamDecoderPtr
-IntegrationUtil::makeSingleRequest(const Network::Address::InstanceConstSharedPtr& addr,
-                                   const std::string& method, const std::string& url,
-                                   const std::string& body, Http::CodecClient::Type type,
-                                   const std::string& host, const std::string& content_type) {
+struct ConnectionCallbacks : public Network::ConnectionCallbacks {
+  ConnectionCallbacks(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+  // Network::ConnectionCallbacks
+  void onEvent(Network::ConnectionEvent event) override {
+    if (event == Network::ConnectionEvent::Connected) {
+      connected_ = true;
+      dispatcher_.exit();
+    } else if (event == Network::ConnectionEvent::RemoteClose) {
+      dispatcher_.exit();
+    } else {
+      if (!connected_) {
+        // Before handshake gets established, any connection failure should exit the loop. I.e. a
+        // QUIC connection may fail of INVALID_VERSION if both this client doesn't support any of
+        // the versions the server advertised before handshake established. In this case the
+        // connection is closed locally and this is in a blocking event loop.
+        dispatcher_.exit();
```

Comment:
I'm unclear on why we need this class.
Does TLS not test connection failures?
Can you snag me on IM and point me at where we're hanging and perhaps I can suggest an alternate fix?

Reply:
can we get a TODO to clean this up?  I think I'd prefer we rework the test framework to handle this failure mode than have this test class. I suspect there's a cleaner way but I don't want to block the PR on it.

Reply:
TODO added.